Arnej/add new encode decode

arnej27959 opened the pull request:
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

@havardpe please review

arnej27959 committed:
add new_encode and new_decode

* move common encode/decode logic from
  simple_tensor.cpp into codec.h / codec.cpp
* encoding and decoding using new Value API
  is in value_codec.h, both for binary format
  and to/from TensorSpec
* will move implementation of TensorSpec
  conversions later
* extend unit test for tensor serialization
  to also test new_encode and new_decode
* add hack in CreateValueFromTensorSpec to
  handle a spec without any cells but
  no mapped dimensions, this was used by
  unit test

arnej27959 committed:
peek into nbostream instead of copying strings

havardpe commented on eval/src/vespa/eval/eval/value_codec.h:
```diff
@@ -0,0 +1,33 @@
+/**
+ * encode a value to binary format
+ **/
+void new_encode(const Value &value, nbostream &output);
```
I suggest we call this encode_value. Also, we could mention that this requires new api compliance in the comment.

havardpe commented on eval/src/vespa/eval/eval/value_codec.h:
```diff
@@ -0,0 +1,33 @@
+/**
+ * decode a value from binary format
+ **/
+std::unique_ptr<Value> new_decode(nbostream &input, const ValueBuilderFactory &factory);
```
I suggest we call this decode_value

havardpe commented on eval/src/vespa/eval/eval/codec.h:
```diff
@@ -0,0 +1,111 @@
+ * dense-subspace consisting of all indexed dimensions that is
+ * uniquely specified by the labels of all mapped dimensions.
+ **/
+struct TypeMeta {
```
this is a vital part of the SimpleTensor implementation that was re-used for encoding purposes, I do not think this should be moved out of like this. I think we should keep this one in SimpleTensor as-is and add a similar DecodePlan to value_codec.cpp

havardpe commented on eval/src/vespa/eval/eval/simple_value.cpp:
```diff
@@ -45,6 +46,10 @@ struct CreateValueFromTensorSpec {
             }
             map[sparse_key][dense_key] = entry.second;
         }
+        // hack for passing some (invalid?) unit tests
```
not really a hack since converting from a spec should use 0.0 as default value for all existing non-defined cells. good catch though.

havardpe commented on eval/src/vespa/eval/eval/simple_value.cpp:
```diff
@@ -45,6 +46,10 @@ struct CreateValueFromTensorSpec {
         }
+        // hack for passing some (invalid?) unit tests
+        if (spec.cells().empty() && type.count_mapped_dimensions() == 0) {
+            map[SparseKey()][0] = 0;
```
prefer using {} or similar instead of SparseKey()

havardpe commented on eval/src/vespa/eval/eval/simple_value.cpp:
```diff
@@ -45,6 +46,10 @@ struct CreateValueFromTensorSpec {
             map[sparse_key][dense_key] = entry.second;
         }
+        // hack for passing some (invalid?) unit tests
+        if (spec.cells().empty() && type.count_mapped_dimensions() == 0) {
```
prefer checking 'map' over spec

havardpe reviewed (commented):
I suggest duplicating the code instead (since we want to remove SimpleTensor alltogether) and put everything from codec.h/cpp into value_codec.cpp (and possibly mutate it to better fit the code)

arnej27959 committed:
revert moving code into codec.h

arnej27959 committed:
move Spec conversion to value_codec.cpp

arnej27959 committed:
cleanup hack

arnej27959 committed:
add unit test for value_codec.h functions

arnej27959 committed:
simplify value_codec some

arnej27959 committed:
refactor without TypeMeta

arnej27959 committed:
use typify_invoke for new_encode also

arnej27959 committed:
cosmetic test fixes

arnej27959 committed:
rename new API

new_encode -> encode_value
new_decode -> decode_value

havardpe commented on eval/src/vespa/eval/eval/value_type.h:
```diff
@@ -60,10 +60,12 @@ class ValueType
     bool is_tensor() const { return (_type == Type::TENSOR); }
     bool is_sparse() const;
     bool is_dense() const;
+    size_t count_indexed_dimensions() const;
```
should add unit test for this

havardpe commented on eval/src/vespa/eval/eval/value_type.h:
```diff
@@ -60,10 +60,12 @@ class ValueType
     size_t dense_subspace_size() const;
     const std::vector<Dimension> &dimensions() const { return _dimensions; }
     std::vector<Dimension> nontrivial_indexed_dimensions() const;
+    std::vector<Dimension> indexed_dimensions() const;
```
should add unit test for this

havardpe commented on eval/src/vespa/eval/eval/value_codec.cpp:
```diff
@@ -0,0 +1,304 @@
+    }
+};
+
+} // namespace <unnamed>
```
should also contain classes below

havardpe reviewed (commented):
some minor comments, otherwise looks very good

arnej27959 committed:
fixes after review

* move namespace end to correct place
* remove ValueType::indexed_dimensions(),
  just loop over all dimensions when getting
  names to avoid making an extra vector
* unit test ValueType::count_indexed_dimensions()
